### PR TITLE
Custom queries

### PR DESCRIPTION
**Pull request by wjohnsto**
Okay, it was not my intent for this to be a big PR, but here it goes:

1. [BH-850 DONE](https://wpengine.atlassian.net/browse/BH-850)
1. [BH-852 DONE](https://wpengine.atlassian.net/browse/BH-852)
1. Queries to get Posts and a Content Node can be customized using fragments and variables
    - This customization can occur with calls to the hooks, services, or during `getServerSideProps` or `getStaticProps`
    - There is a doc now that explains [how queries work](/wpengine/headless-framework/tree/custom-queries/docs/queries)
1. All Next-specific functionality is extracted into a `/packages/headless/src/next` directory
    - You can import it as `import ... from '@wpengine/headless/next';`
1. All React-specific (non-Next) functionality is extracted into a `/packages/headless/src/react` directory
    - You can import it as `import ... from '@wpengine/headless/react';`
1. There is now a generic React `TemplateLoader` that will work outside of Next. There is also a `NextTemplateLoader` that allows Templates to specify `getStaticProps` and `getServerSideProps`
1. Theme components should now export `getServerSideProps` and `getStaticProps` methods if they want to make extra server-side requests [(@see /examples/getting-started/wp-templates/single)](/wpengine/headless-framework/blob/custom-queries/examples/getting-started/wp-templates/single.tsx)
    - Either `getServerSideProps` or `getStaticProps` will be called depending upon whether the app is using `getServerSideProps` or `getStaticProps`
1. Custom queries can be made directly within `getServerSideProps` or `getStaticProps` of pages and theme components as follows:

	```tsx
	import { getApolloClient } from '@wpengine/headless';
	import { gql } from '@apollo/client';
	import { GetStaticPropsContext } from 'next';

	// ...

	export async function getStaticProps(context: GetStaticPropsContext) {
	  const client = getApolloClient(context);
	  client.query({
            query: gql`...`
	  });
	 
	  return getNextStaticProps(context);
	}
	```
	- These queries, if made prior to calling `getNextStaticProps` or `getNextServerSideProps`, will be cached and available client-side
1. There is a generic `authorizeHandler` middleware function that only authorizes a user and sets a cookie. in `@wpengine/headless/next` there is a `previewHandler` that will set preview data and call on the `authorizeHandler`

BREAKING CHANGES:

1. `initializeNextStaticProps` is now `getNextStaticProps` and is in `@wpengine/headless/next`
1. `initializeNextServerSideProps` is now `getNextServerSideProps` and is in `@wpengine/headless/next`
1. `getPropsMiddleware` on Templates is now replaced by `getStaticProps` and `getServerSideProps`
    - The function signature for the new functions is the same as the Next.js variants, so you have access to `context` only (you can import and call for other things you might need)
1. `usePosts` now returns a `WPGraphQL.RootQueryToPostConnection`, so the list of posts will be found on the `nodes` or `edges` properties
1. All Next.js functions are in `@wpengine/headless/next` and not exported from `@wpengine/headless`. This includes:
    -  NextTemplateLoader
    - WPHead
    - getNextServerSideProps
    - getNextStaticPaths
    - getNextStaticProps
    - previewHandler
    - usePost (the Next-specific variant)
    - useUriInfo (the Next-specific variant)
1. All non-Next React functions are in `@wpengine/headless/react` and not exported from `@wpengine/headless`. This includes:
    - HeadlessContext
    - HeadlessProvider
    - Menu
    - MenuList
    - TemplateLoader
    - useApollo
    - useGeneralSettings
    - usePost
    - usePosts
    - useUriInfo
1. The following objects are no longer exported from core:
    - gql (can be imported from `@apollo/client`)
    - useQuery (can be imported from `@apollo/client`)
    - useMutation (can be imported from `@apollo/client`)

### CODE REVIEW

**Comment by mindctrl** [comment]
@wjohnsto woot! Glad to see the ability to customize queries. There's a pretty big merge conflict here now. Can you fix that up before we get too deep in reviewing?

**Review by matt-landers — APPROVED** [review]

**Comment by nickcernis** [comment]
@wjohnsto Thank you for your work on this.

Can you help me to better understand how passing variables to adjust a query should work, please?

I read [the docs](https://github.com/wpengine/headless-framework/tree/custom-queries/docs/queries#customizing-getposts-and-useposts) and amended `packages/headless/src/next/serverSide.ts` like this:

```diff
  if (pageInfo.isPostsPage) {
    await getPosts(client, options.posts);
  }

+  if (pageInfo.isArchive) {
+    await getPosts(client, {
+      variables: {
+        where: { categoryName: 'uncategorized' }, // Hard-coding the category name for now.
+      },
+    });
+  }
```

I expected to see posts from the uncategorized category (default for WP posts) at http://localhost:3000/category/uncategorized. Instead I see no posts. GraphiQL in Chrome shows no `categoryName` variable was passed to `getPosts`.

**Comment by mindctrl** [comment]
I did the same as Nick, but I get different results. At first the category page is empty, and about 3 seconds later all posts are rendered instead of the ones related to the category I requested. 🤷

**Comment by wjohnsto** [comment]
> 
> 
> @wjohnsto Thank you for your work on this.
> 
> Can you help me to better understand how passing variables to adjust a query should work, please?
> 
> I read [the docs](https://github.com/wpengine/headless-framework/tree/custom-queries/docs/queries#customizing-getposts-and-useposts) and amended `packages/headless/src/next/serverSide.ts` like this:
> 
> ```diff
>   if (pageInfo.isPostsPage) {
>     await getPosts(client, options.posts);
>   }
> 
> +  if (pageInfo.isArchive) {
> +    await getPosts(client, {
> +      variables: {
> +        where: { categoryName: 'uncategorized' }, // Hard-coding the category name for now.
> +      },
> +    });
> +  }
> ```
> 
> I expected to see posts from the uncategorized category (default for WP posts) at http://localhost:3000/category/uncategorized. Instead I see no posts. GraphiQL in Chrome shows no `categoryName` variable was passed to `getPosts`.

@nickcernis Can you show me  the contents of the file in question?

**Comment by nickcernis** [comment]
@wjohnsto Added here: https://gist.github.com/nickcernis/8bbc03ecd60cff974435b7bf790dcc72

**Comment by wjohnsto** [comment]
Have you confirmed that the call is being made and the results are missing posts? Something like:

```ts
  if (pageInfo.isArchive) {
    console.log('making call for archive posts')
    const result = await getPosts(client, {
      variables: {
        where: { categoryName: 'uncategorized' }, // Hard-coding the category name for now.
      },
    });
    console.log(result);
  }
```

**Comment by nickcernis** [comment]
> Have you confirmed that the call is being made and the results are missing posts?

Yes, it's firing although the result is not missing posts.

> Something to think about, this same call would need to be made client-side in order to work. You would not be able to call `usePosts()` and expect the archived posts.

Thanks, it sounds like this is the likely cause. I'll look into creating a category template with a matching query.

With that in mind I'm happy to approve and merge and address anything else in follow-ups. Thanks again for your work, @wjohnsto!

**Review by nickcernis — APPROVED** [review]

File: packages/headless/src/api/apolloClient.ts [hunk 1]
```diff
@@ -102,13 +91,20 @@ function createApolloClient(
...
-  initialState = null,
+export function getApolloClient(
+  context?: PersistentContext,
+  initialState: unknown = null,
 ): ApolloClient<NormalizedCacheObject> {
-  const localApolloClient = createApolloClient({
-    cookies: getCookiesFromContext(context),
-  });
+  let localApolloClient:
+    | ApolloClient<NormalizedCacheObject>
```
Comment (mskalandunas) [inline packages/headless/src/api/apolloClient.ts:99]: @wjohnsto hi there, just a random passerby here, but is this a typical union with a prepended `|` or is there something more going on that I may be misunderstanding?

Comment (wjohnsto) [reply packages/headless/src/api/apolloClient.ts:99]: I don't think there is anything more going on, it's possible this was the output from eslint/prettier as I do not remember writing this specifically. The outcome is the same as `let localApolloClient: ApolloClient<NormalizedCacheObject>`.

Comment (mskalandunas) [reply packages/headless/src/api/apolloClient.ts:99]: Awesome, thanks for clearing that up!

**Pull request merged**